src/internal_math.rs

```diff
@@ -1 +1,220 @@
+use std::mem::swap;
 
+/// # Arguments
+/// * `m` `1 <= m`
+///
+/// # Returns
+/// x mod m
+/* const */
+fn safe_mod(mut x: i64, m: i64) -> i64 {
+    x %= m;
+    if x < 0 {
+        x += m;
+    }
+    x
+}
+
+/// Fast moduler by barrett reduction
+/// Reference: https://en.wikipedia.org/wiki/Barrett_reduction
+/// NOTE: reconsider after Ice Lake
+struct Barrett {
+    _m: u32,
+    im: u64,
+}
+
+impl Barrett {
+    /// # Arguments
+    /// * `m` `1 <= m`
+    /// (Note: `m <= 2^31` should also hold, which is undocumented in the original library.
+    /// See the [pull reqeust commment](https://github.com/rust-lang-ja/ac-library-rs/pull/3#discussion_r484661007)
+    /// for more details.)
+    fn new(m: u32) -> Barrett {
+        Barrett {
+            _m: m,
+            im: (-1i64 as u64) / (m as u64) + 1,
+        }
+    }
+
+    /// # Returns
+    /// `m`
+    fn umod(&self) -> u32 {
+        self._m
+    }
+
+    /// # Parameters
+    /// * `a` `0 <= a < m`
+    /// * `b` `0 <= b < m`
+    ///
+    /// # Returns
+    /// a * b % m
+    fn mul(&self, a: u32, b: u32) -> u32 {
+        // [1] m = 1
+        // a = b = im = 0, so okay
+
+        // [2] m >= 2
+        // im = ceil(2^64 / m)
+        // -> im * m = 2^64 + r (0 <= r < m)
+        // let z = a*b = c*m + d (0 <= c, d < m)
+        // a*b * im = (c*m + d) * im = c*(im*m) + d*im = c*2^64 + c*r + d*im
+        // c*r + d*im < m * m + m * im < m * m + 2^64 + m <= 2^64 + m * (m + 1) < 2^64 * 2
+        // ((ab * im) >> 64) == c or c + 1
+        let mut z = a as u64;
+        z *= b as u64;
+        let x = (((z as u128) * (self.im as u128)) >> 64) as u64;
+        let mut v = (z - x * self._m as u64) as u32;
+        if self._m <= v {
+            v = v.wrapping_add(self._m);
```

Comment:
We might want to have test cases that cover this conditional branch.
Thinking about examples...

Reply:
Found an example:
```
m = 2147483647 (= 2^31 - 1)
a = 1073741824 (= 2^30)
b = 2147483646 (= 2^31 - 2)
```
will make `z - x * self._m as u64` overflow.